tracing: Support dynamically loading tracing libraries.

rnburn opened the pull request:
Signed-off-by: Ryan Burn <ryan.burn@gmail.com>

*Description*:

This PR adds support for dynamically loading tracers into Envoy. It will allow authors of OpenTracing compliant tracers to support envoy without requiring envoy to bake their tracers into its build. Tracing libraries can be enabled by adding a section like this to envoy's configuration:
```yaml
  tracing:
    http:
      name: envoy.dynamic.ot
      config:
        library: /path/to/tracing/library
        config: {
          ....
        }
```

With this change, Envoy will be able to use Jaeger's new C++ [tracing library](https://github.com/jaegertracing/cpp-client). Some features this will allow that aren't available when using Jaeger as a collector for Zipkin:
* Propagation will work correctly with other services using Jaeger without requiring them to make [configuration changes](https://github.com/jaegertracing/jaeger-client-go#zipkin-http-b3-compatible-header-propagation).
* Jaeger's different [sampling strategies](https://github.com/jaegertracing/jaeger-client-go#sampling) can be used.
* More efficient [tranports](https://github.com/jaegertracing/jaeger-client-go#span-reporting-transports) are available.

While it might also make sense sometime in the future to offer a version of Jaeger or other tracers built into Envoy, decoupled plug-in support will allow for quicker on-boarding of tracers and more experimentation in the area without making Envoy's external dependencies become unmanageable.

Notes:
* One challenge with using dynamically loaded libraries in Envoy is that the standard c++ library is statically compiled in with the option `-static-libstdc++`. If plugin libraries are linked to the `libstdc++` in a standard way, this will create problems as two versions of the standard library will be loaded and calls to their functions from the plugin will be intermingled between them, breaking many of the standard library functions. The approach I took with Jaeger was to have it also use `-static-libstdc++` and then use an [export map](https://github.com/rnburn/cpp-client/blob/master/export.map) to ensure that it's version of the standard library is used without any intermingling (see [dsohowto](https://www.akkadia.org/drepper/dsohowto.pdf) for background on the approach). It does mean that there will be multiple copies of `libstdc++` running, but from what I've [heard](https://stackoverflow.com/a/47926025/4447365) this can be made to work. Another approach could be to add a config option to link `libstdc++` dynamically.
* This depends on https://github.com/opentracing/opentracing-cpp/pull/45 being merged in to opentracing-cpp, but I wanted to get a PR in early to get feedback on the approach in case we need to make changes to that PR.

*Risk Level*: Medium

*Testing*:
I have a small [example](https://github.com/rnburn/jaeger-envoy-docker/tree/master/example) that demoes this functionality using docker and a branch of [jaeger](https://github.com/rnburn/cpp-client/).

I also plan to add unit test coverage using a mock tracer library that will be added to [opentracing-cpp](https://github.com/opentracing/opentracing-cpp). 

*Docs Changes*:

https://github.com/envoyproxy/data-plane-api/pull/386

*Release Notes*:

Included in PR.

mattklein123 commented:
@rnburn this is sweet. Real quick comment (have not reviewed): This is going to need v2 config defined here: https://github.com/envoyproxy/data-plane-api/blob/master/api/trace.proto. (Everything now is v2 first. v1 is optional). Can you do the data-plane-api PR and then work that into this PR?

mattklein123 commented:
Also, from a quick scan, this will need tests. You are probably going to have to compile some type of  test driver as an .so and load it or something. Not sure what is best.

rnburn commented:
Yes, I'm working a mocktracer library (Similar to what's in [opentracing-go](https://github.com/opentracing/opentracing-go/tree/master/mocktracer)) that will be distributed as part of opentracing-cpp and can be used to test dynamic loading against.

rnburn commented:
I added a data plane [PR](https://github.com/envoyproxy/data-plane-api/pull/386) to support corresponding v2 options and to document the v1 options.

mattklein123 commented:
@rnburn friendly ping on this. If this won't be completed for a while do you mind if we close for now and reopen when ready?

rnburn commented:
Sorry for the delay.

I just added in a mocktracer library implementing the dynamic loading interface to opentracing-cpp (https://github.com/opentracing/opentracing-cpp/commit/359bafe4f5e026400620827c23e8e571ad875c02) that I can use to add testing coverage for the dynamic loading code. I plan to do that this week.

If some blocker comes up and I can't finish the PR soon, I'll close it out and reopen.

rnburn commented:
Updated to use the configuration changes from https://github.com/envoyproxy/data-plane-api/pull/386. I'm adding in testing coverage next.

rnburn commented:
@yurishkuro @isaachier -- are you ok with the build requirements that this method of dynamic loading imposes on the tracing library? Is it a build you could see supporting for Jaeger's C++ client?

Just to summarize: What you need to do is
1) Link dependencies in statically (including the standard C++ library using the flag `-static-libstdc++`).
2) Set up the linker to use an export map so that only Jaeger symbols are exposed.

(I also outlined the approach in this [SO post](https://stackoverflow.com/questions/47841812/how-to-support-dynamic-plugins-when-statically-linking-in-standard-libraries)).

isaachier commented:
I think CMake helps with the exports: https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html. I can fix the linking issue.

mattklein123 commented:
@rnburn re: linking, this is our first foray into dynamic linking which I think we will be doing more of in the future. We can also have a compile option for Envoy which dynamically links if that is preferable for some deployments.

rnburn commented:
So I have a library `mocktracer` that implements the dynamic loading interface and is exposed to [bazel](https://github.com/rnburn/opentracing-cpp/blob/dlopen/mocktracer/BUILD).

I'm trying to figure out how to write a `cc_test` so that `mocktracer` is a dependency that's not linked in but whose location is passed to the unit test so that it can be `dlopen`-ed and tested against.

Any suggestions on how to do this? (also asked on [SO](https://stackoverflow.com/questions/48461242/how-can-you-set-up-dynamically-loaded-library-dependencies-in-bazel]))

rnburn commented:
I found one way to do it: If you list a library in the [data](https://docs.bazel.build/versions/master/be/common-definitions.html#common.data) attribute for cc_test, then it will be available in the current working directory (albeit in a platform-architecture dependent location); but you can then run the `find` command to discover the path... not very elegant, but is there a better way?

mattklein123 commented:
@jmillikin-stripe @htuch any idea on @rnburn question above?

htuch commented:
See https://github.com/envoyproxy/envoy/blob/master/test/test_common/environment.h#L73; this is how you access the data files. Examples are the SSL cert tests, e.g. https://github.com/envoyproxy/envoy/blob/master/test/common/ssl/BUILD#L11 and https://github.com/envoyproxy/envoy/blob/master/test/common/ssl/ssl_socket_test.cc#L1115. In that example, the "rundir" is from a JSON template, but the first link I provided gives you access to the directory as a simple string.

rnburn commented:
The library will show up in `runfilesDirectory` but the path will be something like `_solib_darwin_x86_64/libexternal_Sio_Uopentracing_Ucpp_Smocktracer_Slibmocktracer.so`

Is there an environmental variable for the `_solib_darwin_x86_64` part of the path?

htuch commented:
Maybe a `genrule` to `cp` the source location as given by Bazel into some sane standard location?

jmillikin-stripe commented:
For data files with annoying complex paths like that, I've usually seen them passed to tests with the `args` attribute and a command-line flag. This should work with all test rules:

```
cc_test(
  name = "my_test",
  srcs = ["my_test.cc"],
  data = ["//important:data"],
  args = ["--some-important-data=$(location //important:data)"],
)
```

For C++ tests in particular, I think (based on the docs) you can also put the path into a preprocessor macro with `copts=['-DIMPORTANT_DATA_PATH="$(location //important:data)"']` or `defines=['IMPORTANT_DATA_PATH="$(location //important:data)"']`. I've never tried this though.

mattklein123 commented:
@rnburn as part of this PR can you also please look at fixing coverage in these files? Thanks!
https://s3.amazonaws.com/lyft-envoy/coverage/report-master/coverage.source_common_tracing_opentracing_driver_impl.cc.html
https://s3.amazonaws.com/lyft-envoy/coverage/report-master/coverage.source_common_tracing_lightstep_tracer_impl.cc.html

rnburn commented:
@mattklein123 will do.

rnburn commented:
I'm running into more trouble with bazel and `dlopen`-ing a provided library. If I set up a BUILD file like
```
cc_library(
    name = "b",
    srcs = ["b.cpp"],
    visibility = ["//visibility:public"],
)

cc_library(
    name = "a",
    srcs = ["a.cpp"],
    visibility = ["//visibility:public"],
    deps = ["//:b"]
)

cc_test(
    name = "dlopen_test",
    srcs = ["dlopen_test.cpp"],
    visibility = ["//visibility:public"],
    data = ["//:a"],
)
```
then bazel doesn't handle the dependent library `libb.so` when it sets up `liba.so` for the `dlopen_test`, so running
```
dlopen("/path/to/liba.so", RTLD_NOW | RTLD_LOCAL);
```
from`dlopen_test.cpp`  will fail with
```
dlopen(./_solib_darwin_x86_64/libliba.so, 6): Symbol not found: __Z1gv
  Referenced from: ./_solib_darwin_x86_64/libliba.so
  Expected in: flat namespace
 in ./_solib_darwin_x86_64/libliba.so
```

I wrote a minimal example https://github.com/rnburn/bazel-dlopen to demonstrate the issue.

Anyone know a workaround?

jmillikin-stripe commented:
@rnburn Each `cc_library` links to a separate shared object, which references (but does not include) symbols from its dependencies. If you want a `.so` that contains all symbols from the transitive dependency set, use `cc_binary` like this:

```
cc_binary(
    name = "liba.so",
    srcs = ["a.cpp"],
    linkshared = 1,
    deps = ["//:b"]
)

cc_test(
    name = "dlopen_test",
    srcs = ["dlopen_test.cpp"],
    data = ["//:liba.so"],
)
```

If you want to have both a `cc_library(name="a")` and a `cc_binary(name="liba.so")` in the same BUILD file, you must also tell the library not to create its `.so` (filename conflict). Here's an adjusted version of rnburn/bazel-dlopen/BUILD that passes all tests on my Darwin machine:

```
cc_library(
    name = "b",
    srcs = ["b.cpp"],
)   

cc_library(
    name = "a",
    srcs = ["a.cpp"],
    linkstatic = 1,
    deps = ["//:b"],
)   

cc_binary(
    name = "liba.so",
    linkshared = 1,
    deps = ["//:a"],
)   

cc_test(
    name = "dlopen_test",
    srcs = ["dlopen_test.cpp"],
    data = ["//:liba.so"],
)   

cc_test(
    name = "link_test",
    srcs = ["link_test.cpp"],
    deps = ["//:a"],
)   

```

rnburn commented:
@jmillikin-stripe Looks like that works. Thanks!

htuch commented:
Can you merge master to pickup the latest Bazel changes? I think we might need that to pass on OS X.

rnburn commented:
Undefined behavior sanitizer is giving an error about the function-pointer casting involved when using dlopen/dlsym. I put together a [minimal example](https://gist.github.com/rnburn/e45475ca234ee673380714f5c4e28cf6) of the problem.

I asked on LLVM's IRC channel if there was way to appease the sanitizer, but they didn't have any solution and asked me to put in a bug report. I entered https://github.com/google/sanitizers/issues/911; but in the meantime, I guess I'll just suppress the warnings for that code.

rnburn commented:
@htuch -- done.

mattklein123 commented:
@rnburn is this ready to review again?

rnburn commented:
@mattklein123 not quite yet. I still have some more test cases to write, but getting close.

mattklein123 commented:
No problem. I'm assigning @danielhochman to do first pass on this once it is ready.

rnburn commented:
@danielhochman, @mattklein123 - ok I think this is ready for a first round of reviewing.

danielhochman commented on source/server/config/http/dynamic_opentracing_http_tracer.cc:
```diff
@@ -0,0 +1,38 @@
+  Tracing::DriverPtr dynamic_driver{
+      new Tracing::DynamicOpenTracingDriver{server.stats(), library, config}};
+  return Tracing::HttpTracerPtr(
+      new Tracing::HttpTracerImpl(std::move(dynamic_driver), server.localInfo()));
```
can you use `make_unique` here?

rnburn replied:
Done.

dnoe commented:
@danielhochman is this ready for a senior maintainer to take a final pass?

danielhochman commented on source/server/configuration_impl.h:
```diff
@@ -54,6 +54,11 @@ class HttpTracerFactory {
+  /**
+   * Returns true if the tracing driver requires cluster name to be defined.
+   */
+  virtual bool requiresClusterName() { return true; }
```
this can be a const method

rnburn replied:
Fixed.

danielhochman commented on bazel/repository_locations.bzl:
```diff
@@ -34,8 +34,8 @@ REPOSITORY_LOCATIONS = dict(
-        commit = "e57161e2a4bd1f9d3a8d3edf23185f033bb45f17",
-        remote = "https://github.com/opentracing/opentracing-cpp", # v1.2.0
+        commit = "1368b7de0343388a0e217643481ec4d475ed22e5",
+        remote = "https://github.com/rnburn/opentracing-cpp",
```
do you know when https://github.com/opentracing/opentracing-cpp/pull/45 will merge?

rnburn replied:
I'm hoping to get it merged in this week.

alyssawilk commented on source/common/tracing/dynamic_opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,38 @@
+      opentracing::DynamicallyLoadTracingLibrary(library.c_str(), error_message);
+  if (!library_handle_maybe) {
+    if (error_message.empty()) {
+      throw EnvoyException{fmt::format("{}", library_handle_maybe.error().message())};
```
Is it possible to test these cases, here and below?

rnburn replied:
I simplified so that there's only a single code path now that should get hit in the testing.

alyssawilk commented on source/server/config/http/dynamic_opentracing_http_tracer.h:
```diff
@@ -0,0 +1,29 @@
+
+  std::string name() override;
+
+  bool requiresClusterName() override { return false; }
```
Seems to be unused/ untested?

rnburn replied:
Added test coverage.

rnburn commented:
@alyssawilk - An example would be helpful. I have a branch of Jaeger's tracer that I've been testing with; but it's not available in a release yet. Also, not sure how to include an example with envoy since it requires Jaeger's C++ library to be built.

https://www.envoyproxy.io/docs/envoy/v1.5.0/install/sandboxes/jaeger_tracing.html shows how to set up tracing to a Jaeger collector by using its Zipkin compatibility feature. But Jaeger uses a different default propagation [format](https://github.com/jaegertracing/jaeger-client-go#zipkin-http-b3-compatible-header-propagation) so that has the disadvantage that it won't interoperate well with other services instrumented with Jaeger.

Also, using Jaeger's native tracing library will bring other advantages such as sampling support, more efficient transports, and key-value logging (if envoy ever decides to use that).

alyssawilk commented:
Hm, how about a TODO to land a walkthrough of how to use native tracing once the version of Jaeger you're using lands?  I don't think we need docs for every tracer open-tracing allows, and I don't think actually doccing things up needs to block this PR, but I think at least one example would be super helpful.

Other than that and waiting for your open-tracing PR to land I think you're good to go!

alyssawilk commented on source/common/tracing/dynamic_opentracing_driver_impl.cc:
```diff
@@ -12,23 +12,15 @@ DynamicOpenTracingDriver::DynamicOpenTracingDriver(Stats::Store& stats, const st
-    } else {
-      throw EnvoyException{fmt::format("{}: {}", tracer_maybe.error().message(), error_message)};
-    }
+    throw EnvoyException{fmt::format("{}: {}", tracer_maybe.error().message(), error_message)};
```
I definitely thing this is easier to read than the prior code, but won't it be a bit confusing if error_message is empty?  Is that an unexpected corner case?

rnburn replied:
It would have a extra trailing ":" if the error message is empty which might be a little confusing.

I could add something like a formatErrorMessage function that could do the conditional formatting and be tested directly if you think that's better?

alyssawilk replied:
If it's likely to be empty I'd make it conditional.  Knowing little of opentracing I'm not sure if it's warranted. :-)

rnburn replied:
Ok. Put the conditional formatting back in, but used a separate function so that everything has test coverage.

rnburn commented:
I added a TODO for an example;  I merged in https://github.com/opentracing/opentracing-cpp/pull/45 and I updated the build so that it no longer points to my fork of opentracing-cpp.

I'm going to put in a few more commits to address https://github.com/envoyproxy/envoy/pull/2252#issuecomment-362023791; then I think I should be caught up with everything.

alyssawilk commented:
code LGTM modulo the two build failures.  Clean those up and you're good to go

rnburn commented:
Ok. I fixed the build. I also added more test coverage so that now these files from my last PR are covered:
https://28730-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_tracing_opentracing_driver_impl.cc.html
https://28730-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_tracing_lightstep_tracer_impl.cc.html

htuch commented on source/common/tracing/dynamic_opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,38 @@
+    throw EnvoyException{formatErrorMessage(tracer_maybe.error(), error_message)};
+  }
+  tracer_ = std::move(*tracer_maybe);
+  RELEASE_ASSERT(tracer_ != nullptr);
```
Why is this `RELEASE_ASSERT` and not `ASSERT`?

htuch commented on source/common/tracing/dynamic_opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,38 @@
+  RELEASE_ASSERT(tracer_ != nullptr);
+}
+
+std::string DynamicOpenTracingDriver::formatErrorMessage(std::error_code error_code,
```
This should probably be in the anonymous namespace, it's not using any state from the instance.

htuch commented on source/common/tracing/dynamic_opentracing_driver_impl.cc:
```diff
@@ -0,0 +1,38 @@
+
+DynamicOpenTracingDriver::DynamicOpenTracingDriver(Stats::Store& stats, const std::string& library,
+                                                   const std::string& tracer_config)
+    : OpenTracingDriver{stats} {
```
Tiny nit: FYI, most of Envoy doesn't use brace style when calling super.
